File: src/main/java/ast/ASTFileLocation.java
package ast;

import java.io.*;

import compiler.Main;
import org.antlr.v4.runtime.*;

public class ASTFileLocation {
    public long line_num;
    public long line_pos;
    public String fileName;

    public ASTFileLocation(long line, long pos, String fileName){
        line_num = line;
        line_pos = pos;
        this.fileName = fileName;
    }

    public static ASTFileLocation fromToken(Token n){
        var l = new ASTFileLocation(n.getLine(), n.getCharPositionInLine(), Main.fileName);
        return l;
    }

    public void print(){
        try {
            var file = new BufferedReader(new FileReader(fileName));
            for(var i = 0; i < line_num-1; i++){
                file.readLine();
            }
            String line = file.readLine();
            System.out.println(line);

            for(var i = 0; i < line_pos; i++){
                if(line.charAt(i) == '\t'){
                    System.out.print("\t");
                } else {
                    System.out.print(" ");
                }
            }
            System.out.println("^");


        } catch (IOException err){
            System.out.println("IOException");
        }

    }
}

File: src/main/kotlin/ast/ASTFileLocation.kt
package ast

import java.io.*

import compiler.Main
import org.antlr.v4.runtime.*

class ASTFileLocation(val line_num: Int, val line_pos: Int, val fileName: String) {

    fun print() {
        try {
            val file = BufferedReader(FileReader(fileName))
            for (i in 0 until line_num - 1) {
                file.readLine()
            }
            val line = file.readLine()
            System.out.println(line)

            for (i in 0 until line_pos) {
                if (line[i] == '\t') {
                    System.out.print("\t")
                } else {
                    System.out.print(" ")
                }
            }
            System.out.println("^")


        } catch (err: IOException) {
            System.out.println("IOException")
        }

    }

    companion object {

        fun fromToken(n: Token): ASTFileLocation {
            return ASTFileLocation(n.line.toInt(), n.charPositionInLine.toInt(), Main.fileName)
        }
    }
}
